gemd/entity/object/process_spec.py

```diff
@@ -56,31 +57,53 @@ class ProcessSpec(BaseObject, HasParameters, HasConditions, HasTemplate):
     def __init__(self, name, *, template=None,
                  parameters=None, conditions=None,
                  uids=None, tags=None, notes=None, file_links=None):
+        from gemd.entity.object.ingredient_spec import IngredientSpec
+        from gemd.entity.link_by_uid import LinkByUID
+
         BaseObject.__init__(self, name=name, uids=uids, tags=tags, notes=notes,
                             file_links=file_links)
         HasParameters.__init__(self, parameters=parameters)
         HasConditions.__init__(self, conditions=conditions)
-
-        self._ingredients = []
+        HasTemplate.__init__(self, template=template)
 
         # By default, a ProcessSpec is not linked to any MaterialSpec.
         # If a MaterialSpec is linked to this ProcessSpec,
         # then the field self._output_material will be automatically populated
         self._output_material = None
-
-        HasTemplate.__init__(self, template=template)
+        self._ingredients = validate_list(None, [IngredientSpec, LinkByUID])
 
     @property
     def ingredients(self):
         """Get the list of input ingredient specs."""
         return self._ingredients
 
-    def _unset_ingredient(self, ingred):
-        """Remove `ingred` from this process's list of ingredients."""
-        if ingred in self._ingredients:
-            self._ingredients.remove(ingred)
-
     @property
     def output_material(self):
         """Get the output material spec."""
         return self._output_material
+
+    # To avoid infinite recursion, fast return on revisit
+    global eq_seen
+    eq_seen = set()
+
+    def __eq__(self, other):
+        if (self, other) in eq_seen:  # Cycle encountered
+            return True  # This will functionally be & with the correct result of ==
```

Comment:
Does this get lit up in the tests?

Reply:
Yes.  When the Process Spec iterates over its Ingredient Specs, each `ingredient.process` bounces through this check.